File: backend/app/regime/indicators.py
"""Vectorized technical indicators used for regime detection and scoring features.

All functions take a DataFrame with columns open/high/low/close(/volume) indexed
by time (ascending) and return a pandas Series aligned to that index.
"""
from __future__ import annotations

import numpy as np
import pandas as pd


def true_range(df: pd.DataFrame) -> pd.Series:
    prev_close = df["close"].shift(1)
    ranges = pd.concat(
        [df["high"] - df["low"], (df["high"] - prev_close).abs(), (df["low"] - prev_close).abs()], axis=1
    )
    return ranges.max(axis=1)


def atr(df: pd.DataFrame, period: int = 14) -> pd.Series:
    return true_range(df).ewm(alpha=1 / period, adjust=False, min_periods=period).mean()


def adx(df: pd.DataFrame, period: int = 14) -> pd.Series:
    up_move = df["high"].diff()
    down_move = -df["low"].diff()

    plus_dm = np.where((up_move > down_move) & (up_move > 0), up_move, 0.0)
    minus_dm = np.where((down_move > up_move) & (down_move > 0), down_move, 0.0)

    tr_smooth = true_range(df).ewm(alpha=1 / period, adjust=False, min_periods=period).mean()
    plus_di = 100 * pd.Series(plus_dm, index=df.index).ewm(alpha=1 / period, adjust=False, min_periods=period).mean() / tr_smooth
    minus_di = 100 * pd.Series(minus_dm, index=df.index).ewm(alpha=1 / period, adjust=False, min_periods=period).mean() / tr_smooth

    dx = 100 * (plus_di - minus_di).abs() / (plus_di + minus_di).replace(0, np.nan)
    return dx.ewm(alpha=1 / period, adjust=False, min_periods=period).mean()


def bollinger_bandwidth(df: pd.DataFrame, period: int = 20, num_std: float = 2.0) -> pd.Series:
    sma = df["close"].rolling(period).mean()
    std = df["close"].rolling(period).std(ddof=0)
    upper = sma + num_std * std
    lower = sma - num_std * std
    return (upper - lower) / sma.replace(0, np.nan)


def choppiness_index(df: pd.DataFrame, period: int = 14) -> pd.Series:
    tr_sum = true_range(df).rolling(period).sum()
    high_max = df["high"].rolling(period).max()
    low_min = df["low"].rolling(period).min()
    span = (high_max - low_min).replace(0, np.nan)
    return 100 * np.log10(tr_sum / span) / np.log10(period)


def trend_slope_r2(df: pd.DataFrame, period: int = 20) -> tuple[pd.Series, pd.Series]:
    """Rolling linear-regression slope (in price units/bar) and R^2 of `close`."""
    close = df["close"]
    x = np.arange(period)
    x_mean = x.mean()
    x_var = ((x - x_mean) ** 2).sum()

    def _slope_r2(window: np.ndarray) -> tuple[float, float]:
        y = window
        y_mean = y.mean()
        cov = ((x - x_mean) * (y - y_mean)).sum()
        slope = cov / x_var if x_var else 0.0
        pred = slope * (x - x_mean) + y_mean
        ss_res = ((y - pred) ** 2).sum()
        ss_tot = ((y - y_mean) ** 2).sum()
        r2 = 1 - ss_res / ss_tot if ss_tot else 0.0
        return slope, r2

    slopes = np.full(len(close), np.nan)
    r2s = np.full(len(close), np.nan)
    values = close.to_numpy()
    for i in range(period - 1, len(values)):
        s, r = _slope_r2(values[i - period + 1 : i + 1])
        slopes[i] = s
        r2s[i] = r
    return pd.Series(slopes, index=df.index), pd.Series(r2s, index=df.index)


def atr_percentile(df: pd.DataFrame, atr_period: int = 14, lookback: int = 100) -> pd.Series:
    atr_series = atr(df, atr_period)
    return atr_series.rolling(lookback, min_periods=max(10, lookback // 4)).rank(pct=True)


def realized_vol_zscore(df: pd.DataFrame, ret_period: int = 1, lookback: int = 100) -> pd.Series:
    returns = df["close"].pct_change(ret_period)
    rolling_mean = returns.rolling(lookback).mean()
    rolling_std = returns.rolling(lookback).std(ddof=0)
    return (returns - rolling_mean) / rolling_std.replace(0, np.nan)

File: backend/src/regime/indicators.ts
/**
 * Technical indicators used for regime detection and scoring features.
 * Operates on plain ascending-time OHLCV bar arrays; each function returns an
 * array aligned to the input (NaN where there isn't enough warm-up data yet).
 */
export interface OhlcBar {
  time: Date;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

// A single degenerate bar (e.g. a bad data-feed tick briefly producing a
// wildly wrong OHLC bar) otherwise dominates every ATR-based calculation
// downstream for many minutes afterward -- Wilder's smoothing only decays by
// ~(1-1/period) per bar, so one huge true-range value stays influential long
// after the underlying bad tick is corrected. 2026-07-21 incident: a single
// zero-volume NQ bar with a false ~29,000-point range (and, via prevClose,
// the very next bar too) kept a live continuous-scan signal's ATR-derived
// stop distance in the thousands of points for several minutes after the
// bad tick itself had already self-healed. Clip each bar's raw true range to
// at most this multiple of the recent local median -- generous enough that
// a real volatility spike (a genuine gap, a news event) still reads as
// elevated, tight enough that one degenerate bar can't dominate the average.
const TRUE_RANGE_OUTLIER_CLIP_MULTIPLE = 10;
const TRUE_RANGE_OUTLIER_CLIP_WINDOW = 14;

function clipTrueRangeOutliers(raw: number[]): number[] {
  return raw.map((value, i) => {
    const reference = raw.slice(Math.max(0, i - TRUE_RANGE_OUTLIER_CLIP_WINDOW), i); // preceding bars only, excludes this one
    if (reference.length < Math.max(3, Math.floor(TRUE_RANGE_OUTLIER_CLIP_WINDOW / 2))) return value; // not enough history yet to judge
    const sorted = [...reference].sort((a, b) => a - b);
    const median = sorted[Math.floor(sorted.length / 2)]!;
    if (median <= 0) return value;
    return Math.min(value, median * TRUE_RANGE_OUTLIER_CLIP_MULTIPLE);
  });
}

export function trueRange(bars: OhlcBar[]): number[] {
  const raw = bars.map((bar, i) => {
    if (i === 0) return bar.high - bar.low;
    const prevClose = bars[i - 1]!.close;
    return Math.max(bar.high - bar.low, Math.abs(bar.high - prevClose), Math.abs(bar.low - prevClose));
  });
  return clipTrueRangeOutliers(raw);
}

/** Wilder's smoothing (equivalent to pandas ewm(alpha=1/period, adjust=False)). */
function wilderSmooth(values: number[], period: number): number[] {
  const out = new Array<number>(values.length).fill(NaN);
  const alpha = 1 / period;
  let prev: number | undefined;
  for (let i = 0; i < values.length; i++) {
    if (i < period - 1) continue;
    if (prev === undefined) {
      prev = mean(values.slice(0, i + 1));
    } else {
      prev = alpha * values[i]! + (1 - alpha) * prev;
    }
    out[i] = prev;
  }
  return out;
}

function mean(xs: number[]): number {
  return xs.reduce((a, b) => a + b, 0) / xs.length;
}

export function atr(bars: OhlcBar[], period = 14): number[] {
  return wilderSmooth(trueRange(bars), period);
}

export function adx(bars: OhlcBar[], period = 14): number[] {
  const plusDm: number[] = [0];
  const minusDm: number[] = [0];
  for (let i = 1; i < bars.length; i++) {
    const upMove = bars[i]!.high - bars[i - 1]!.high;
    const downMove = bars[i - 1]!.low - bars[i]!.low;
    plusDm.push(upMove > downMove && upMove > 0 ? upMove : 0);
    minusDm.push(downMove > upMove && downMove > 0 ? downMove : 0);
  }

  const trSmooth = wilderSmooth(trueRange(bars), period);
  const plusDiSmooth = wilderSmooth(plusDm, period);
  const minusDiSmooth = wilderSmooth(minusDm, period);

  const dx: number[] = bars.map((_, i) => {
    const tr = trSmooth[i]!;
    const plusDi = (100 * plusDiSmooth[i]!) / tr;
    const minusDi = (100 * minusDiSmooth[i]!) / tr;
    const denom = plusDi + minusDi;
    return denom === 0 ? NaN : (100 * Math.abs(plusDi - minusDi)) / denom;
  });

  return wilderSmooth(
    dx.map((v) => (Number.isNaN(v) ? 0 : v)),
    period
  );
}

export function bollingerBandwidth(bars: OhlcBar[], period = 20, numStd = 2): number[] {
  const closes = bars.map((b) => b.close);
  return closes.map((_, i) => {
    if (i < period - 1) return NaN;
    const window = closes.slice(i - period + 1, i + 1);
    const sma = mean(window);
    const std = Math.sqrt(mean(window.map((v) => (v - sma) ** 2)));
    if (sma === 0) return NaN;
    return (2 * numStd * std) / sma;
  });
}

export function choppinessIndex(bars: OhlcBar[], period = 14): number[] {
  const tr = trueRange(bars);
  return bars.map((_, i) => {
    if (i < period - 1) return NaN;
    const trWindow = tr.slice(i - period + 1, i + 1);
    const highWindow = bars.slice(i - period + 1, i + 1).map((b) => b.high);
    const lowWindow = bars.slice(i - period + 1, i + 1).map((b) => b.low);
    const trSum = trWindow.reduce((a, b) => a + b, 0);
    const span = Math.max(...highWindow) - Math.min(...lowWindow);
    if (span <= 0) return NaN;
    return (100 * Math.log10(trSum / span)) / Math.log10(period);
  });
}

/** Rolling linear-regression slope (price units/bar) and R^2 of `close`. */
export function trendSlopeR2(bars: OhlcBar[], period = 20): { slope: number[]; r2: number[] } {
  const closes = bars.map((b) => b.close);
  const slope = new Array<number>(bars.length).fill(NaN);
  const r2 = new Array<number>(bars.length).fill(NaN);

  const x = Array.from({ length: period }, (_, i) => i);
  const xMean = mean(x);
  const xVar = x.reduce((acc, v) => acc + (v - xMean) ** 2, 0);

  for (let i = period - 1; i < closes.length; i++) {
    const y = closes.slice(i - period + 1, i + 1);
    const yMean = mean(y);
    const cov = x.reduce((acc, xv, idx) => acc + (xv - xMean) * (y[idx]! - yMean), 0);
    const s = xVar ? cov / xVar : 0;
    const pred = x.map((xv) => s * (xv - xMean) + yMean);
    const ssRes = y.reduce((acc, yv, idx) => acc + (yv - pred[idx]!) ** 2, 0);
    const ssTot = y.reduce((acc, yv) => acc + (yv - yMean) ** 2, 0);
    slope[i] = s;
    r2[i] = ssTot ? 1 - ssRes / ssTot : 0;
  }
  return { slope, r2 };
}

/** Percentile rank (0-1) of the latest ATR value within a rolling lookback window. */
export function atrPercentile(bars: OhlcBar[], atrPeriod = 14, lookback = 100): number[] {
  const atrValues = atr(bars, atrPeriod);
  return atrValues.map((_, i) => {
    const minPeriods = Math.max(10, Math.floor(lookback / 4));
    const start = Math.max(0, i - lookback + 1);
    const window = atrValues.slice(start, i + 1).filter((v) => !Number.isNaN(v));
    if (window.length < minPeriods || Number.isNaN(atrValues[i]!)) return NaN;
    const current = atrValues[i]!;
    const rank = window.filter((v) => v <= current).length;
    return rank / window.length;
  });
}

export function realizedVolZscore(bars: OhlcBar[], lookback = 100): number[] {
  const closes = bars.map((b) => b.close);
  const returns = closes.map((c, i) => (i === 0 ? NaN : (c - closes[i - 1]!) / closes[i - 1]!));
  return returns.map((_, i) => {
    if (i < lookback) return NaN;
    const window = returns.slice(i - lookback + 1, i + 1).filter((v) => !Number.isNaN(v));
    const m = mean(window);
    const std = Math.sqrt(mean(window.map((v) => (v - m) ** 2)));
    if (!std) return NaN;
    return (returns[i]! - m) / std;
  });
}

export function lastValid(values: number[]): number | null {
  for (let i = values.length - 1; i >= 0; i--) {
    if (!Number.isNaN(values[i]!)) return values[i]!;
  }
  return null;
}
